### ISSUE TITLE
core variance enforcement bug

### ISSUE THREAD

**Issue by joshuareagan** [urbit/arvo#891]
tl;dr -- how rigorously should we enforce the core variance rules?  Certainly more strictly than we are doing it now, but by how much?  Discuss.

Right now core variance isn't correctly enforced by the type system.  @ohAitch gives one example of that here: https://github.com/urbit/arvo/issues/457

The problem specifically is this: there are rules associated with each type of core metal about how the payload can be used, and those rules aren't enforced.  Gold cores (the normal kind) are supposed to have a red-write payload; no problem there.  Iron cores should have a write-only sample and an opaque context---no reading or writing to the context, where "writing" here means using wing syntax to produce a mutant copy of the core, e.g., wing(old-val new-val).  Zinc cores have a read-only sample and an opaque context.  Read-only means that you shouldn't be able to use zinc gates in function calls (among other things).  Anton's example above exploits the fact we don't enforce the read-only nature of zinc samples.  Lead cores have an opaque payload -- no reading or writing anything in the payload.

Right now you can read or write in places you aren't supposed to.

I've been working on this issue lately and I anticipate that at least a partial fix is forthcoming.  In the potential fix given now, the "read" restrictions are only partially enforced.  Name resolution to the payload results in a "find-name" error.  But lark/direct address syntax allows you to get a noun fragment from an address with all type info thrown away (i.e., the fragment is always type *).   This is a compromise of what it means to be "read-only".

What do you guys think -- is this bad?

@cgyarvin is inclined to think so.  He leans toward absolute enforcement, i.e., forbidding lark syntax workarounds.  But the problem with enforcing reads this strictly is that we have a few commonly used gates whose functionality depends on being able to read the sample of an iron gate.  I have in mind `++corl` and `++curr` of `hoon.hoon`.  (The same is true of `++cury` but it's easy to fix it not to need to read a sample.)  `corl` and `curr` are quite cleverly written, but I don't see any obvious way to redo them without in some way reading from an iron sample.  

Thoughts?  @ohAitch do you have any ideas?

**Comment by cgyarvin** [comment]
Why are we using iron samples here?  Is there any actual reason to do so?

**Comment by ohAitch** [comment]
Throwing away type information seems correct: the reason for variance is
that a core could be "subclassed" with an incompatible type, e.g. a  |+(*
...)  passed into a sample of type  |+(^ ...)  , such that its +< is not a
^ type. Removing all type guarantees for the read solves that issue.

(Also in the limit case, ofc you can just +<:`*`the-core, or .*(the-core [0
6]) if the compiler is trying to be particularly clever)

**Comment by ohAitch** [comment]
Agree with Curtis that iron samples to higher-order wet gates are generally
questionable, since you're already getting your variance through the
wetness mechanism, and allowing full type read access allows you to provide
a more "polite" sample wrt enabling composition. (Which is what the
cleverness in cork is trying to do, iirc)

**Comment by cgyarvin** [comment]
1. Yep that’s exactly what I meant.
2. Always tighten as much as possible.  Easy to loosen.

**Comment by joshuareagan** [comment]
Ok, this is straightforward at the urbit layer.  I may need help next week to figure out how to get the jets to cooperate...

**Comment by joshuareagan** [comment]
I'm missing something here.  I modified `++corl` to the following (no other changes to vanilla release-candidate):

```
++  corl                                                ::  compose backwards
  |*  {a/_|=(* +<) b/_|=(* +<)}
  =<  +:|.((a (b)))      ::  type check
  =+  c=+<.b
  |@  ++  $  (a (b c))
  --
```

When I try doing two resets on this modification, I get a bail:

```
[%vega-start-hoon /~zod/home/~2018.11.17..04.18.44..2cdc/sys/hoon]
%vega-parsed
%vega-compiled
[%vega-arvo /~zod/home/~2018.11.17..04.18.44..2cdc/sys/arvo]
%vega-minted
%arvo-assembly
%arvo-assembled
[%vega-kernel 0x4b03.f66b]
[%vega-compiled 141 141]
%hoon-load
[%tang /~zod/home/~2018.11.17..04.18.44..2cdc/sys/zuse ~sardef-dornyl]
[%vane %a /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/ames ~dospen-bilmud]
[%vane %b /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/behn ~locmex-fonber]
[%vane %c /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/clay ~torben-sonsug]
[%vane %d /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/dill ~ladrem-nodfyl]
[%vane %e /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/eyre ~tidbep-ronlyr]
[%vane %f /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/ford ~lardex-bidwes]
[%vane %g /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/gall ~nammus-pildun]
[%vane %j /~zod/home/~2018.11.17..04.18.44..2cdc/sys/vane/jael ~mastes-fabsev]
~zod:dojo> |reset|
bail: foul
bailing out
Assertion failed: (0), function u3m_bail, file ../noun/manage.c, line 653.
                                                                          Abort trap: 6 (core dumped)
```

I thought this would work.  What am I missing?

**Comment by cgyarvin** [comment]
A foul always means some kind of corruption...

**Comment by joshuareagan** [comment]
This was a totally fresh fakezod, with newly cloned arvo and urbit from the release-candidate branch...

**Comment by ohAitch** [comment]
the * cast completely defeats the point of getting at the sample, at that
point you can just =| c=* and ignore the sample completely?

**Comment by joshuareagan** [comment]
Actually the cast is probably a bad idea -- presumably we want the composed gate to keep its sample type.  I'm back to square one...

**Comment by ohAitch** [comment]
Square one being "gold gates work, iron gates mostly work"? :P

(the sample type in turn only matters if you use it directly, by bunting or
like in another corl; just calling the composed gate regularly will swap it
out before doing the |@ wet recompilation)

**Comment by ohAitch** [comment]
++roll is probably a more central example than "another corl"

**Comment by joshuareagan** [comment]
Curtis wants a 100% ban on reading iron samples, even with lark syntax.  (1) That breaks `+corl` as it is currently written, because it reads the sample.  (2) We could cast only for gold cores, but then `+corl` doesn't compose iron gates, which seems bad.  (3) We could force `+corl` to have a bunted noun sample, but that restricts the kind of gates that can be composed.  I don't have a sense of whether there are `corl`ed gates with cell samples.

If we only do a soft ban on reading iron cores with lark syntax, `+corl` works as-is even after the changes.  With a strict ban I can't think of a solution that handles all three concerns above.  (That doesn't mean there isn't one...)

**Comment by joshuareagan** [comment]
I guess doing a `*` cast lets us recover the correct structure of the sample of the gate to be composed, even if not its type.  And because the composed gate created by `+corl` is wet, we don't care about the sample type as much as we do about the raw noun there.  So now I think the cast solution is okay again.  Does that make sense?

**Comment by joshuareagan** [comment]
Nvm, the cast solution doesn't work for `+roll` because it does a cast using an iron gate sample type...  I should have thought a little more about your response @ohAitch !

**Comment by ohAitch** [comment]
oh, if you do a hard ban that will make a bunch of higher-order functions
more clunky yeah. Which I suppose is not inconsistent with Curtis's general
opinions thereupon :P

(specifically, each one has to choose to either not work with iron gates,
or turn gold gate input into effectively iron gate outputs, with the
current soft-ban sample-* behavior)

I don't think there's much you can do with a raw noun with a removed type?
You can't bunt the resulting corl gate for example, even with gold inputs.
For a simplified example, consider (|*(f=gate |*(a=_+<.f (f a))) |=(a=@ a))
vs (|*(f=gate |*(a=_+<:`*`f (f a))) |=(a=@ a))

**Comment by joshuareagan** [comment]
Yes, this seems right to me.  As far as I can tell the hard ban effectively makes it impossible for a function to (1) accept both gold and iron gates and (2) make any sort of use of the input gate sample type.

The soft ban is interesting in the case of a wet gate, because the lark syntax work-around can be used to work with the run-time input gate sample type.  I don't really see this option as a breach of type safety, though I'd have to think about that more to be sure...

**Comment by ohAitch** [comment]
to be overly pedantic, "run-time" is misleading, but "per-call-site"
certainly.

It is definitely not a breach of type safety: any core will ~(repo ut ...)
to just %noun, which means inserting effectively the ^- * iff necessary is
consistent with how the rest of the system thinks about cores.

**Comment by joshuareagan** [comment]
It's not overly pedantic imo, that's a good correction.

As far as type safety goes, I had in mind the use of genericity (wetness) with iron gate sample types. Right now we use this combo all over the place, for good reason.  If the soft-ban option is to be a real option, we need to be totally certain that no type inference funny-business happens in the soft-ban semi-loophole.

**Comment by cgyarvin** [comment]
As long as Josh is confident in type safety, he did the work so it’s his call.

**Comment by ohAitch** [comment]
I might be a broken record, but: the only thing even no-ban leaks is type
information. * is the absence of type information.

(for example, it leaks ? in place of @ for ^-($-(? ~) |=(a=@ ~)), if the
iron sample type is taken as too canonical)

**Comment by joshuareagan** [comment]
@ohAitch your last response is a little too cryptic for me to follow.  Right now we effectively have a soft-ban in place for reads (soft-ban = reads on restricted areas using lark gets you a raw noun).  My proposed changes would leave that in place, but actually enforce the writing restrictions.

A genuine no-ban would be bad and would break type safety -- leaking type info is enough to do this.

A proper variance scheme allows us to pass gates only if the input value type contravaries with respect to the gate type.  I.e., if we do things correctly, one gate A nests under another gate B iff the sample of B nests under the sample of A.

So your example makes sense -- an atom is broader than a boolean, so the contravariance of the sample type is respected.  (Perhaps would have been a reason not to make the booleans atoms, but we *did* make them atoms.  And since we did, your snippet is legitimate.)

For others who may not quite follow -- let's say I'm writing a function F that takes some gate G.  Let's also say I want G to be able to take as input any *mammal*.  In other words, F is going to pass arbitrary *mammal*s to G, so it needs to know how to handle all *mammal*s correctly.  We can't pass F a gate that only takes *dogs* as input, because F might call it with a *cat*.  But F can take a gate that takes all *animal*s as input, because a gate that can handle an *animal* can handle any *mammal*, no problem.

What's important is that we isolate F from inferring any type from G's sample.  After G is passed in, we want to infer only that G's sample is *mammal* -- we know that's all we're going to call it with in F.  Hence the read-only restriction -- it's to forbid the possibility of errant type inference based on what G can do in the wild with its broader sample.

So, in fact, Curtis's instinct to want to do a strict ban on reading iron samples is sound.  We don't want to do any type inference based on the fact that the sample can be read as a noun.  If we only dealt with dry gates I would agree with Curtis 100% and implement the strict ban without complaint.

But when we bring genericity into the picture, we do want to do type inference using the original iron gate sample type, because we potentially want to cast for that type as we do in `+roll`.   Right now we exploit the soft read ban and use lark syntax to implement this kind of genericity when passing gates.  I can't think of a better way to do this with a hard read ban... but we still want to be %100 sure a soft ban respects type safety before making that call if it's what we do.

**Comment by joshuareagan** [comment]
Sorry, I wasn't totally clear on how reading iron samples breaks type safety -- I want F to do a type check to make sure it only passes *mammal*s to G, even if G can actually handle all *animal*s.  If I screw up F and leave open the possibility of passing a *lizard* to G, then I deserve a nest-fail.  The restriction on reading the iron sample makes sure I get what I deserve.

**Comment by ohAitch** [comment]
Hm, so this sounds like the reading implementation is actually completely
fine, and the problem is writing(which yes, it is a contradiction to have a
zinc/lead gate, covariance is for cores). Are there examples similar to
#457 of /reading/ producing type errors, with master hoon.hoon?

**Comment by joshuareagan** [comment]
The current implementation (soft-ban) seems mostly fine to me.  The lark syntax to get a noun is sort of cheating though.  That's because, ideally, you should be able to, within F, read G's sample and be guaranteed to get a *mammal*.  But you aren't, because its default sample may be a *lizard*.  Certainly most nouns aren't *mammal*s either , so allowing any read at all is a genuine compromise.  But if that's the cost of making full use of genericity with lead gate samples I think we could do a lot worse.

Zinc and lead gates can actually make sense, as long as you can't do function calls on them (or analogous de-sugared $ resolutions with changes).  Imagine using a zinc gate as an object in object-oriented programming -- there are definitely cases in what you'd want the sample (i.e. state + ID info on the kind of object) to co-vary with the gate type.

**Comment by joshuareagan** [comment]
That should read "But if that's the cost of making full use of genericity with *iron* gate samples I think we could do a lot worse."

**Comment by ohAitch** [comment]
I think "you aren't guaranteed to get a *mammal*" is precisely what
contravariance means, actually :)

**Comment by joshuareagan** [comment]
Well, sure, that's right.  But the point of the contravariance restrictions is that, if strictly enforced, we can safely sustain the illusion that G's sample is a *mammal*.  This is mostly but not absolutely 100% true if we compromise with lark.

**Comment by joshuareagan** [comment]
Another day, another variance bug.  This time it's with zinc cores... the sample nesting rules don't seem to be enforced correctly, unless I'm mistaken.

When casting for a zinc core of type `X`, I produce a zinc core of type `Y`.  The sample of `Y` doesn't nest under the sample of `X`, yet Hoon allows for `Y` to nest under `X`.  This shouldn't be!

**Comment by cgyarvin** [comment]
Can you state this more clearly with an example?

**Comment by joshuareagan** [comment]
Examples to illustrate:

```
> =zinc-core ^&  |=(^ 15)

> `_zinc-core`|=([@ @] 15)
< 1&xqz
  { {* *}
    {our/@p now/@da eny/@uvJ}
    <19.rcz 24.vrg 5.kkw 36.apb 119.smn 240.mmm 51.zox 93.glm 74.dbd 1.qct $141>
  }
>

> `_zinc-core`|=(* 15)
< 1&xqz
  { {* *}
    {our/@p now/@da eny/@uvJ}
    <19.rcz 24.vrg 5.kkw 36.apb 119.smn 240.mmm 51.zox 93.glm 74.dbd 1.qct $141>
  }
>
```

The first one goes through correctly.  The second one incorrectly goes through.  Strangely the following correctly nest-fails:

```
> ^+(^&(|=(^ 15)) |=(* 16))
nest-fail
```

**Comment by cgyarvin** [comment]
Strange indeed. Sounds like more heroic debugging, but hopefully not on thanksgiving!

**Comment by cgyarvin** [comment]
Can you actually use this to create a type violation, eg, treating a cell as an atom?

**Comment by joshuareagan** [comment]
How about this:

```
> =sum |_  [a=@ b=@]
  ++  $  (add a b)
  ++  alchemy  .
  --

> (sum 12 13)
25

> =a 22

> =b 33

> =prototype ^&  |_  ?([@ @] @)
  ++  $  (add a b)
  ++  alchemy  .
  --

> =unsafe |=(a=_prototype ~&(%runtime (alchemy.a 15)))

%runtime
> (unsafe sum)
ford: %ride failed to execute:
```

**Comment by cgyarvin** [comment]
When I look at this example, what jumps out at me is that these two cores shouldn't be compatible.  This should be a nest-fail.  Reason: +alchemy in both cores produces, of course, a gold core, and these gold cores should not be compatible.

You can write a +alchemy line on any core.  That arm will produce that core as a gold core.  But you're not creating a type hole, or shouldn't be -- you're just producing a core that can't be used polymorphically.  For two cores to match, the arms of the subtype must match the arms of the base type, and their products must nest within the products of the base type.  But the entire reason we're using variance is that these cores, if gold, will be incompatible.  So this test should produce a mismatch on the +alchemy arm.

I suspect the problem is that you didn't update the logic in =nest to match your call site transformation?

**Comment by joshuareagan** [comment]
Your response seems right -- I *am* getting a nest-fail on a master branch arvo, for example, so the problem here has something to do with my changes.  I hadn't realized that `+nest` depended on the old `+fond` location of gold transformations for correct type inference.

Your comments *also* extend to the previous "bug" I mentioned above -- I introduced that too.

**Comment by joshuareagan** [comment]
I spoke too soon!  These bugs have nothing to do with the location of the gold transformation, whether +fond or +fire.

Both of these bugs are present in a fresh release candidate which don't have my changes.  So they were introduced sometime between the current master and the current release-candidate.

**Comment by joshuareagan** [comment]
I'm putting this here for me as much as anyone else, but I found another problem.  This nest goes through when it shouldn't:

```
> =a 12

> =core1 ^?  |=(b=@ a)

> =a "hello"

> =core2 ^?  |=(b=@ a)

> ^+(core1 core2)
< 1?ojs
  { b/@
    {a/@ud our/@p now/@da eny/@uvJ}
    <19.oht 24.kqz 5.dtg 36.apb 119.dzc 240.mmm 51.mcd 93.eta 74.dbd 1.qct $141>
  }
>
```
This should fail because the $ arm in the latter core produces a tape, which shouldn't nest under the type of the $ arm in the former core, which produces a @ ud.

I have a pretty good guess about what's causing this problem though.  (2nd line of the %core case in +dext of +nest).  I'll work on this tomorrow with @benjamin-tlon.

**Comment by cgyarvin** [comment]
Honestly, man, I always knew our type system was overdue for some real
adversarial testing.  All the holes that people normally stumble through
are plugged, and that's about it.

**Comment by ohAitch** [comment]
!

And this is release-candidate, cc-release, master? I feel like I've ever
gotten a |? cast fail, which this suggests wouldn't happen.

**Comment by joshuareagan** [comment]
This is all on release-candidate.  Strangely, that last example I gave correctly nest-fails on master.

It looks like master, while buggy in certain respects, is a bit tighter than release-candidate on variance.

**Comment by joshuareagan** [comment]
Another variance bug, left here for note-taking -- wings with arm names can be used to evade variance restrictions.

This wing resolution to a lead sample fails, which is the desired outcome:

```
> =b ^?  |=(sample=@ 15)

> sample.b
-find.sample.b
```

But if we use the `$` arm in the wing it goes through (unsafe):

```
> sample.$.b
0

> ? sample.$.b
  @
0
```

**Comment by cgyarvin** [comment]
Yeah that’s a true bug.  We’re not actually running an arm but it thinks we are.

**Comment by joshuareagan** [comment]
Working on the last bug now.  (The others are resolved!)  It's easy to fix at the Hoon layer, but I may need a little assistance to fix the corresponding jet.  Afaict the problem is `u3qfu_felt`.

In Hoon, `+felt` takes an `opal` and produces a `type`.  `+felt` is used by `+fond` for adjusting the subject recursively as each element of the wing is resolved.  The problem at the Hoon layer is that `+felt` uses `+fire` when the input `opal` is for an arm; the `+fire` product is the desired `type` to be returned by `+felt`.  That's fine except `+fire` turns cores gold, which we don't want for subject-adjustment in `+fond`.  So I added a boolean test to `+fire` which allows one to run it without alchemy and adjusted `+felt` accordingly.  Easy.

But `u3qfu_felt` doesn't work like `+fond`: there's no call to the jet for `+fire`, and none of the desired checks that makes `+fire` enforce type safety.  I'm not totally clear on what `u3i_molt` and `usj_cook` do, so any insight on those is welcome.  Other than that, I'm thinking about how to make `u3qfu_felt` do what the corrected `+felt` does.

**Comment by joemfb** [comment]
`u3i_molt` is like `%=`, it takes a noun a zero-terminated varargs list of axis/noun pairs. `u3j_cook` is a hook, a way to call back into a core from a jet.

**Comment by joshuareagan** [comment]
Okay, I think I follow.  Thanks!

**Comment by joshuareagan** [comment]
Bugs in this thread have been resolved:
https://github.com/urbit/urbit/pull/1089
https://github.com/urbit/urbit/pull/1078
https://github.com/urbit/urbit/pull/1079
https://github.com/urbit/arvo/pull/914
https://github.com/urbit/arvo/pull/902
https://github.com/urbit/arvo/pull/871
